ArduPlane/mode_qloiter.cpp

```diff
@@ -3,12 +3,121 @@
+        plane.quadplane.loiter_nav->clear_pilot_desired_acceleration();
+        plane.quadplane.loiter_nav->init_target();
+        return;
+    }
+    if (!plane.quadplane.motors->armed()) {
+        _enter();
+    }
+
+    plane.quadplane.check_attitude_relax();
+
+    if (plane.quadplane.should_relax()) {
+        plane.quadplane.loiter_nav->soften_for_landing();
+    }
+
+    const uint32_t now = AP_HAL::millis();
+    if (now - plane.quadplane.last_loiter_ms > 500) {
+        plane.quadplane.loiter_nav->clear_pilot_desired_acceleration();
+        plane.quadplane.loiter_nav->init_target();
+    }
+    plane.quadplane.last_loiter_ms = now;
+
+    // motors use full range
+    plane.quadplane.set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
+
+    // set vertical speed and acceleration limits
+    plane.quadplane.pos_control->set_max_speed_accel_z(-plane.quadplane.get_pilot_velocity_z_max_dn(), plane.quadplane.pilot_velocity_z_max_up, plane.quadplane.pilot_accel_z);
+
+    // process pilot's roll and pitch input
+    float target_roll_cd, target_pitch_cd;
+    plane.quadplane.get_pilot_desired_lean_angles(target_roll_cd, target_pitch_cd, plane.quadplane.loiter_nav->get_angle_max_cd(), plane.quadplane.attitude_control->get_althold_lean_angle_max());
+    plane.quadplane.loiter_nav->set_pilot_desired_acceleration(target_roll_cd, target_pitch_cd);
+    
+    // run loiter controller
+    if (!plane.quadplane.pos_control->is_active_xy()) {
+        plane.quadplane.pos_control->init_xy_controller();
+    }
+    plane.quadplane.loiter_nav->update();
+
+    // nav roll and pitch are controller by loiter controller
+    plane.nav_roll_cd = plane.quadplane.loiter_nav->get_roll();
+    plane.nav_pitch_cd = plane.quadplane.loiter_nav->get_pitch();
+
+    if (now - plane.quadplane.last_pidz_init_ms < (uint32_t)plane.quadplane.transition_time_ms*2 && !plane.quadplane.is_tailsitter()) {
+        // we limit pitch during initial transition
+        float pitch_limit_cd = linear_interpolate(plane.quadplane.loiter_initial_pitch_cd, plane.quadplane.aparm.angle_max,
+                                                  now,
+                                                  plane.quadplane.last_pidz_init_ms, plane.quadplane.last_pidz_init_ms+plane.quadplane.transition_time_ms*2);
+        if (plane.nav_pitch_cd > pitch_limit_cd) {
+            plane.nav_pitch_cd = pitch_limit_cd;
+            plane.quadplane.pos_control->set_externally_limited_xy();
+        }
+    }
+    
+    
+    // call attitude controller with conservative smoothing gain of 4.0f
+    plane.quadplane.attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(plane.nav_roll_cd,
+                                                                                  plane.nav_pitch_cd,
+                                                                                  plane.quadplane.get_desired_yaw_rate_cds());
+
+    if (plane.control_mode == &plane.mode_qland) {
```

Comment:
If this function is moved to mode.cpp then this becomes some sort of overridable method instead - `run_z` or something - QLAND and guided doing different things to most.